### ISSUE TITLE
LSTM class_weight exception

### ISSUE THREAD

**Issue by MaxHoefl** [fchollet/keras#3440]
Hi,

my problem is related to [#1624](https://github.com/fchollet/keras/issues/1624).

I have the following model for a binary classification task.
![lstm_diagram](https://cloud.githubusercontent.com/assets/14946739/17552433/e6e61354-5ef7-11e6-9917-67cf02c3dcf8.png)

The dataset has N observations of 40-dimensional inputs. They are first fed into 120-dim LSTM layer which in turn passes its hidden units on to several TimeDistributedDense layers until it reaches a sigmoid layer. 
I present the training data X_train as N x 1 x 40 dimensional array and the labels (0/1) as 
N x 1 x 1 array. This is my code (the dropout layers were neglected in the chart above):

```
model = Sequential()
model.add(LSTM(120, input_shape=(None, X_train.shape[2]), 
return_sequences=True, consume_less='gpu'))
model.add(TimeDistributed(Dense(50, activation='relu')))
model.add(Dropout(0.2))
model.add(TimeDistributed(Dense(20, activation='relu')))
model.add(Dropout(0.2))
model.add(TimeDistributed(Dense(10, activation='relu')))
model.add(Dropout(0.2))
model.add(TimeDistributed(Dense(3, activation='relu')))
model.add(TimeDistributed(Dense(1, activation='sigmoid')))
```

As the labels are highly imbalanced, I passed class weights to fit
`model.fit(X_train, y_train, nb_epoch=num_epoch, batch_size=batch_size, verbose=0, shuffle=False, callbacks=[epochPrint], class_weight=class_weights)`
where `class_weights={0: 1, 1: 25}`.

**I am getting an error:** `class_weight not supported for 3+ dimensional targets`.

Now @fchollet answered to the issues stated above that "The concept of 'class' is ill-defined is that case (3D labels)". But when I present the labels as a N x 1 array, e.g. [0,0,1,1,0,0,...], I get the exception `Error when checking model target: expected timedistributed_5 to have 3 dimensions, but got array with shape (5932720, 1)`. 

I have a bad feeling that I misunderstood how the LSTM layer works. I appreciate any help!

Best regards,
Max

**Comment by dolaameng** [comment]
Have you tried using parameter `sample_weight` with shape `(samples, sequence_length)`, and specify `sample_weight_mode="temporal"` in `compile()`, as suggested in the [document](https://keras.io/models/sequential/)?

**Comment by MaxHoefl** [comment]
No not yet, will try that. I just noticed that I am totally off anyways. When I am presenting the data as `(nb_samples, 1, nb_features)` doesn't that defy the whole point of LSTM being able to pass on states between time steps? As in the image above I rather want say `nb_timesteps=10` and then construct inputs by rolling a window across the original time series.

There is a pretty cool alternative to `sample_weight` suggested in [#2115](https://github.com/fchollet/keras/issues/2115):
Say we had 2 classes and we want to penalize the model stronger when it predicts a false negative, i.e. y_pred = 0 and y_true = 1. Then one could use the following loss in `model.compile()`
```
def w_categorical_crossentropy(y_true, y_pred, weights):
    nb_cl = len(weights)
    final_mask = K.zeros_like(y_pred[:, 0])
    y_pred_max = K.max(y_pred, axis=1)
    y_pred_max = K.reshape(y_pred_max, (K.shape(y_pred)[0], 1))
    y_pred_max_mat = K.equal(y_pred, y_pred_max)
    for c_p, c_t in product(range(nb_cl), range(nb_cl)):
        final_mask += (weights[c_t, c_p] * y_pred_max_mat[:, c_p] * y_true[:, c_t])
    return K.categorical_crossentropy(y_pred, y_true) * final_mask

w_array = np.ones((2,2))
w_array[1, 7] = 1.2
w_array[7, 1] = 1.2
ncce = partial(w_categorical_crossentropy, weights=w_array)
```

One question regarding `sample_weight` though. Say I wanted to use it in a many-to-one model like such 
![screen shot 2016-08-11 at 11 11 07](https://cloud.githubusercontent.com/assets/14946739/17585559/6a120aa0-5fb4-11e6-9bda-06ef417315f8.png)
So I would write something like 
```
model = Sequential()
model.add(LSTM(120, input_shape=(nb_timesteps, nb_features)))
model.add(Dense(1,activation='sigmoid'))
```
right? So I am passing in as inputs `N` windows of size `(nb_timesteps, nb_features)`. Does that mean my `sample_weight` has to be of shape `(N, nb_timesteps)` ? Could I use in the many-to-one case the `class_weight` parameter again?

Thanks a lot for your help,
Max

**Comment by dolaameng** [comment]
Yes I believe the shape of the `sample_weight` should be `(N, nb_timesteps)` in your case if your LSTM returns a sequence and your last Dense is TimeDistributed. It gives the way to specify for both each sample and each time step.

If the model doesn't return a sequence, the sample weights should be a 1D array as long as nsamples.
